### PR TITLE
adding related concept model

### LINKED ISSUE #8: Concept relationships should be done at concept level
At the moment concept relationships are defined in English term for compatibility with IEV termbuilder (easier transition). This needs to be fixed though.

### LINKED ISSUE #33: Implement RelatedConcept model

### PR DESCRIPTION
**Pull request by HassanAkbar**
- Implemented `RelatedConcept` model.

closes #33 
closes #8

### CODE REVIEW

File: lib/glossarist/concept.rb [hunk 1]
```diff
@@ -73,6 +75,14 @@ def default_designation
       localized&.terms&.first&.designation
     end
 
+    # All Related Concepts
+    # @return [Array<RelatedConcept>]
+    attr_reader :related
+
+    def related=(related)
+      @related = related.map { |r| RelatedConcept.new(r) }
```
Comment (ronaldtse) [inline lib/glossarist/concept.rb:83]: What happens if the object passed in here is already a RelatedConcept object? Preferably the new just returns the object itself.

File: lib/glossarist/related_concept.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    TYPES = %i[
+      deprecates
+      supersedes
+      narrower
+      broader
+      equivalent
+      compare
+      contrast
+      see
+    ]
```
Comment (ronaldtse) [inline lib/glossarist/related_concept.rb:16]: I want to be sure that we can traverse both ways, ie. Supersede, and superseded by. A relationship contains a subject, object and a relation type.

**Review by ronaldtse — CHANGES_REQUESTED** [review]

**Comment by ronaldtse** [comment]
Thanks @HassanAkbar !

**Comment by ronaldtse** [comment]
@HassanAkbar can you please incorporate #25 into this PR so that we don't miss functionality? Thanks.

File: lib/glossarist/concept.rb (hunk 1 above)
Comment (HassanAkbar) [reply lib/glossarist/concept.rb:117]: I've updated the `new` method to return the object if it is already initialized.

File: lib/glossarist/concept.rb [hunk 3]
```diff
@@ -15,8 +15,42 @@ class Concept < Model
     # @return [Hash<String, LocalizedConcept>]
     attr_reader :localizations
 
+    def deprecated_concepts
```
Comment (ronaldtse) [inline lib/glossarist/concept.rb:18]: 1. Can we have a simple `related` concept relation?
2. For these `def deprecated_concepts`, the `deprecated` part is clearly derived from the relation name. Can we generate these methods? We should use the same verb form, i.e. `deprecates_concepts` for `:deprecates`

File: lib/glossarist/concept.rb [hunk 4]
```diff
@@ -42,7 +76,7 @@ def to_h
       {
         "termid" => id,
         "term" => default_designation,
-        "related" => related_concepts,
+        "related" => related&.map(&:to_h),
```
Comment (ronaldtse) [inline lib/glossarist/concept.rb:79]: There will be a problem when related concept is related to itself, i.e. circular relation. In the related concepts, we should likely only provide their unique identifiers.

In addition, we will need to serialise a Glossarist Term into JSON-LD (in the near future), so imagine using GraphQL to retrieve related concepts.

**Review by ronaldtse — CHANGES_REQUESTED** [review]

File: lib/glossarist/concept.rb (hunk 3 above)
Comment (HassanAkbar) [reply lib/glossarist/concept.rb:18]: > 1. Can we have a simple related concept relation?

We do have a simple `related concept` relation.

> For these `def deprecated_concepts`, the `deprecated` part is clearly derived from the relation name. Can we generate these methods? We should use the same verb form, i.e. `deprecates_concepts` for `:deprecates`

From what I understand you want to change the method names to `<relation_type>_concepts` i.e `deprecates_concepts`, `supersedes_concepts`, `compare_concepts` etc.

@ronaldtse did I get this right?

File: lib/glossarist/concept.rb (hunk 4 above)
Comment (HassanAkbar) [reply lib/glossarist/concept.rb:79]: `related` is an array containing the `RelatedConcept` objects. It only contains a reference to the related object. the schema for RelatedConcept is

```JSON
{
    "type": "<type of relation>",
    "content": "<String>",
    "ref": "<Ref>",
}
```

And Ref contains
```JSON
{
    "source": "<String>",
    "id": "<String>",
    "version": "<String>",
}
```

According to this schema, the reference will only contain the identifier.

File: lib/glossarist/concept.rb (hunk 3 above)
Comment (ronaldtse) [reply lib/glossarist/concept.rb:18]: Yes, meta programming would make things cleaner.

File: lib/glossarist/concept.rb (hunk 4 above)
Comment (ronaldtse) [reply lib/glossarist/concept.rb:79]: OK thanks!

**Review by ronaldtse — APPROVED** [review]
Thanks @HassanAkbar , can you rebase and merge?

**Pull request merged**